cxx-qt-gen/src/parser/qobject.rs

```diff
@@ -0,0 +1,479 @@
+// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
+// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
+//
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use crate::syntax::{
+    attribute::{attribute_find_path, attribute_tokens_to_ident},
+    path::{path_angled_args_to_type_path, path_compare_str, path_to_single_ident},
+};
+use quote::format_ident;
+use std::collections::HashMap;
+use syn::{
+    spanned::Spanned, Error, Ident, ImplItem, Item, ItemEnum, ItemImpl, ItemStruct, Result, Type,
+    TypePath,
+};
+
+#[derive(Default)]
+pub struct ParserCxxQtData {
+    /// Map of the QObjects defined in the module that need generation
+    pub qobjects: HashMap<Ident, ParserQObject>,
+    /// Any use blocks end up in the CXX-Qt generated module
+    pub uses: Vec<Item>,
+}
+
+/// A representation of a QObject within a CXX-Qt ItemMod
+///
+/// This has initial splitting of Item's into relevant blocks, other phases will
+/// then mutate these Item's for generation purposes.
+#[derive(Default)]
+pub struct ParserQObject {
+    /// This is the Data struct that currently stores the properties for the QObject
+    ///
+    /// In the future this will be removed
+    pub data_struct: Option<ItemStruct>,
+    /// This is the RustObj struct that stores the invokables for the QObject
+    pub rust_struct: Option<ItemStruct>,
+    /// This is the Signals enum that defines the Q_SIGNALS for the QObject
+    pub signal_enum: Option<ItemEnum>,
+    /// This is the list of methods that need to be implemented on the C++ object in Rust
+    ///
+    /// Ones marked as invokable will need to be exposed to C++ as Q_INVOKABLE
+    pub methods: Vec<ImplItem>,
+    /// This is update request handler for the QObject
+    ///
+    /// In the future this may be removed
+    pub update_requester_handler: Option<ItemImpl>,
+    /// Items that we don't need to generate anything for CXX or C++
+    /// eg impls on the Rust object or Default implementations
+    pub others: Vec<Item>,
+}
+
+/// Find the QObject's within the module
+///
+/// Returns a hashmap with the ident of the QObject as the key and an empty ParserQObject as the value
```

Comment:
Returning empty ParserQObjects seems odd... I can tell from the code that uses this that the empty ParserQObjects are populated by `find_and_merge_cxx_qt_item`, but somehow it feels strange that `find_qobjects` and `find_and_merge_cxx_qt_item` are free functions with completely different names that operate on the same data structure. Just looking at the documentation of these functions and the function signatures, I can't tell there's any relationship between the functions without seeing how they're used in context.

Reply:
I've changed these to be methods on the `ParsedCxxQtData` struct. So there is `find_qobject_keys` that searches for the qobject's and adds them to the hashmap. Then `parse_cxx_qt_item` that adds any CXX-Qt items to the qobjects or returns them if they are others.

Reply:
I think that's a good solution.